webapp/tests/Utils/Scoreboard/ScoreboardIntegrationTest.php

```diff
@@ -0,0 +1,281 @@
+            [ 'team' => $this->teams[0], 'rank' => 2, 'solved' => 0, 'time' => 0 ],
+            [ 'team' => $this->teams[1], 'rank' => 1, 'solved' => 2, 'time' => 161 ],
+        ];
+
+        // Jury scoreboard should not depend on freeze, so test a couple.
+        foreach ([ '+0:30:00', '+1:00:00', '+1:20:00' ] as $freeze) {
+            $this->contest->setFreezetimeString($freeze);
+            $this->recalcScoreCaches();
+
+            $scoreboard = $this->ss->getScoreboard($this->contest, true);
+            $this->assertScoresMatch($expected_scores, $scoreboard->getScores());
+        }
+    }
+
+    public function testScoreboardPublicFreeze()
+    {
+        $this->contest->setFreezetimeString('+1:10:00');
+        $this->createDefaultSubmissions();
+        $this->recalcScoreCaches();
+
+        $expected_scores = [
+            [ 'team' => $this->teams[0], 'rank' => 2, 'solved' => 0, 'time' => 0 ],
+            [ 'team' => $this->teams[1], 'rank' => 1, 'solved' => 1, 'time' => 69 ],
+        ];
+
+        $scoreboard = $this->ss->getScoreboard($this->contest, false);
+        $this->assertScoresMatch($expected_scores, $scoreboard->getScores());
+    }
+
+    function assertScoresMatch($expected_scores, $scores)
+    {
+        foreach ( $expected_scores as $row ) {
+            $team = $row['team'];
+            $name = $team->getName();
+
+            $score = $scores[$team->getTeamid()];
+            $this->assertInstanceOf(TeamScore::class, $score);
+
+            $this->assertEquals($row['rank'],   $score->getRank(), "Rank for '$name'");
+            $this->assertEquals($row['solved'], $score->getNumberOfPoints(), "# solved for '$name'");
+            $this->assertEquals($row['time'],   $score->getTotalTime(), "Total time for '$name'");
+        }
+    }
+
+    function recalcScoreCaches()
+    {
+        $this->em->flush();
+        $this->ss->refreshCache($this->contest);
+    }
+
+    function createDefaultSubmissions()
+    {
+        $lang = $this->em->getRepository(Language::class)->find('cpp');
+
+        $team = $this->teams[0];
+        $this->createSubmission($lang, $this->problems[0], $team, 53*60+15.053, 'no-output');
+        $this->createSubmission($lang, $this->problems[0], $team, 53*60+57.240, null);
+
+        $team = $this->teams[1];
+        $this->createSubmission($lang, $this->problems[0], $team, 69*60+00.000, 'correct');
+        $this->createSubmission($lang, $this->problems[0], $team, 72*60+07.824, 'wrong-answer');
+        $this->createSubmission($lang, $this->problems[1], $team, 72*60+39.733, 'wrong-answer');
+        $this->createSubmission($lang, $this->problems[1], $team, 72*60+59.999, 'correct');
+        $this->createSubmission($lang, $this->problems[1], $team, 79*60+00.000, 'wrong-answer');
+        $this->createSubmission($lang, $this->problems[2], $team, 84*60+42, null);
+        $this->createSubmission($lang, $this->problems[2], $team, 85*60+42, 'correct')
+            ->setValid(false);
+    }
```

Comment:
can you please add an explicit comment here that the 3rd team did not submit anything?

Reply:
Done.